### PR TITLE
Switch between maps at run time

### LINKED ISSUE #1089: Specific Tilemap Saving/Loading
## Summary

Implement the ability to specify the map being saved/loaded.

## Goal 

Currently when you save a map, you can not name it. So each map you save gets automatic successive naming (UnnamedMap.json, UnnamedMap_02.json, etc). Also, when you load a map it does not give you an option to choose what map to load, it just automatically chooses the latest (i think) map you saved.

Ideally it would be nice to be able to name a tilemap when saving and/or select from the list of saved tilemaps so you can choose one to overwrite. And when loading a tilemap, it should display a list of the saved tilemaps for you to choose one to load.

## Document

Current documentation on Tilemaps can be found here: https://ss3d.gitbook.io/systems/tilemaps

### PR DESCRIPTION
**Pull request by stilnat**
# Summary

Add the possibility to save and load different tilemaps, naming and renaming them.

#### PR checklist
- [x] The game builds properly without errors.

- [x] No unrelated changes are present.

- [x] No "trash" files are committed.

- [x] Relevant code is documented.

- [ ] Update the related GitBook document, or create a new one if needed.

# Pictures/Videos)

# Testing

#### Networking checklist

- [x] Works from host in host mode.

- [x] Works from server in server mode.

- [x] Works on server in client mode.

- [x] Works and is syncronized across different clients.

- [x] Is persistent.

# Changes

Cut the tilemenu script into smaller pieces, one script for each tab, handling displaying and clearing their content.
Did a lot of changes on the menu prefab. Simplified it a lot but also changed how it looks a lot.

# Related issues/PRs 

close #1089

# TODO

- [x] Separate code in multiple script, one for each tab (build, remove, save, load)
- [x] Add a ITileMenuTab interface with method refresh and method show, implemented by all scripts above.
- [x] Refresh content upon clicking a tab.
- [x] Do not change map name if a map is already named like the string entered
- [x] Add renaming map.
- [x] Maybe inherit from input field so they automatically handle toggling off controls (because they should all do that and there's multiple)
- [x] Add an enum, for all tabs in the menu, so TileMapMenu knows which tab to refresh and load when clicking on a tab, and which one is currently showing.
- [x] integrate search bar, layer selection dropdown and build or delete button into the content area of the build tab.
- [x] Capture input when renaming.
- [x] Change display of renaming input while renaming.
- [ ] try to fix the "commit" input issue of the renaming when the controls are staying off.

# Known issue

- pressing enter when changing the name of a map does not restore control to player. I can't figure out which event should trigger when pressing enter, I've tried a lot of different things. I posted an question here. https://forum.unity.com/threads/which-event-or-method-is-raised-in-tmp-input-fields-when-pressing-enter.1524796/

### CODE REVIEW

**Review by cosmiccoincidence — DISMISSED** [review]
Change "Delete" text in build menu to purple to match hologram color.

Is the client supposed to be able to change map too? Only seemed to work on host.

There's some small bugs it creates but are the sort of thing that can be figured out in later issues.

Will want someone with some code experience to look it over though. And will need documentation updated and linked before merge.

**Comment by stilnat** [comment]
@cosmiccoincidence will change the color. The client was not able to load the map before. I guess we can add later that functionnality if we need to. Won't be too hard, just a call to the server and checking rights of the client.

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapBuildTab.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            if (IsDeleting)
+            {
+                tmpComponent.text = " Delete";
+                tmpComponent.color = new Color(0xA9, 0x00, 0xFF, 0xFF);
+            }
+            else
+            {
+                tmpComponent.text = " Build";
+                tmpComponent.color = new Color(0x31, 0x32, 0xDD, 0xFF);
+            }
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapBuildTab.cs]: These color values are cryptic. Add a const field for them and name them something readable like "BuildColorRed"

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapCreator.cs [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+using UnityEngine.InputSystem;
+using UnityEngine.UIElements;
+using InputSystem = SS3D.Systems.Inputs.InputSystem;
+
+namespace SS3D.Systems.Tile.TileMapCreator
+{
+    /// <summary>
+    /// In-game editor for placing and deleting items/objects in a tilemap.
+    /// </summary>
+  /*  public class TileMapCreator : NetworkSystem, IPointerEnterHandler, IPointerExitHandler
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapCreator.cs]: If a class is unused or outdated, you can either delete it or add the [Obsolete] tag and add the reasoning

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                // I've no idea why but something modify the slots scale so it's necessary to adjust it here.
+                slot.transform.localScale= Vector3.one;
+
+                MapNameSlot mapNameSlot = slot.GetComponent<MapNameSlot>();
+                mapNameSlot.MapNameField.readOnly = true;
+                mapNameSlot.MapNameField.text = mapNameWithNoExtension;
+
+                mapNameSlot.MapNameField.onSelect.AddListener(delegate { LoadMap(mapNameWithNoExtension); });
+                mapNameSlot.DeleteButton.onClick.AddListener(() => DeleteMap(mapNameWithNoExtension));
+                mapNameSlot.DeleteButton.onClick.AddListener(() => Refresh());
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs]: Not sure how I feel about adding two methods in this button, I'd join them.

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                GameObject slot = Instantiate(_mapNameSlotPrefab, _loadMapContentRoot.transform, true);
+
+                // I've no idea why but something modify the slots scale so it's necessary to adjust it here.
+                slot.transform.localScale= Vector3.one;
+
+                MapNameSlot mapNameSlot = slot.GetComponent<MapNameSlot>();
+                mapNameSlot.MapNameField.readOnly = true;
+                mapNameSlot.MapNameField.text = mapNameWithNoExtension;
+
+                mapNameSlot.MapNameField.onSelect.AddListener(delegate { LoadMap(mapNameWithNoExtension); });
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs]: Make this a function, not a delegate

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        /// </summary>
+        /// <param name="mapNameField"> The input field containing the map to rename.</param>
+        private void Rename(ControlsOffInputField mapNameField)
+        {
+            string oldName = mapNameField.text;
+            mapNameField.readOnly = false;
+            mapNameField.onSelect.RemoveAllListeners();
+            mapNameField.ActivateInputField();
+            mapNameField.onEndEdit.AddListener(delegate { RenameSave(mapNameField, oldName);});
+            mapNameField.onDeselect.AddListener(delegate { RenameSave(mapNameField, oldName); });
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs]: Move to function and not delegate

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            }
+            else
+            {
+                string savePath = Subsystems.Get<TileSystem>().SavePath;
+                SaveSystem.RenameFile(savePath + "/" + oldName, savePath + "/" + mapNameField.text);
+            }
+
+            Refresh();
+
+            mapNameField.onSelect.AddListener(delegate { LoadMap(mapNameField.text); });
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs:154]: move to function

File: Assets/Scripts/SS3D/Systems/Tile/TileSystem.cs [hunk 7]
```diff
@@ -127,13 +130,13 @@ public bool CanBuild(TileObjectSo tileObjectSo, Vector3 placePosition, Direction
         }
 
         [Server]
-        public void Save()
+        public void Save(string mapName, bool overwrite)
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Tile/TileSystem.cs:133]: override, not overwrite

**Review by joaoburatto — CHANGES_REQUESTED** [review]

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapBuildTab.cs (hunk 1 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapBuildTab.cs]: done

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapCreator.cs (hunk 2 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapCreator.cs]: done

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs (hunk 3 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs]: done

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs (hunk 4 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs]: done

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs (hunk 5 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs]: done

File: Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs (hunk 6 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapLoadTab.cs:154]: done

File: Assets/Scripts/SS3D/Systems/Tile/TileSystem.cs (hunk 7 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Tile/TileSystem.cs:133]: [overwrite : to replace information in (a computer file) with new information. overwrite the existing file.](https://www.britannica.com/dictionary/overwrite)

also override is a c# key word

**Review by joaoburatto — APPROVED** [review]

**Pull request merged**